main/guides/js-programming/hardened-js.md

```diff
@@ -0,0 +1,388 @@
+with [Airbnb JavaScript Style](https://github.com/airbnb/javascript#variables--unary-increment-decrement), we recommend:
+ - Avoid `--` and `++` operators.
+ - Break lines after `{` and before `}`.
+
+## Objects and the _maker_ pattern
+
+Let's unpack the `makeCounter` example a bit.
+
+JavaScript is somewhat novel in that objects need not belong to any
+class; they can just stand on their own:
+
+<<< @/snippets/test-hardened-js.js#singleton
+
+We can make a new such object each time a function is called
+using the _maker pattern_:
+
+<<< @/snippets/test-hardened-js.js#maker
+
+::: tip Use lexically scoped variables rather than properties of this.
+The style above avoids boilerplate such as `this.x = x; this.y = y`.
+:::
+
+::: tip Use arrow functions
+We recommend [arrow function](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions)
+syntax rather than `function makePoint(x, y) { ... }` declarations
+for conciseness and to avoid `this`.
+:::
+
+## Defensive objects with `harden()`
+
+By default, anyone can clobber the properties of
+our objects so that they fail to conform to the expected API:
+
+<<< @/snippets/test-hardened-js.js#clobber
+
+Worse yet is to clobber a property so that it misbehaves but
+covers its tracks so that we don't notice:
+
+<<< @/snippets/test-hardened-js.js#exploit
+
+Our goal is **defensive correctness**: a program is _defensively correct_ if it remains correct despite arbitrary behavior on the part of its clients. _For further discussion, see [Concurrency Among Strangers](http://erights.org/talks/promises/paper/tgc05.pdf) and other [Agoric papers on Robust Composition](https://papers.agoric.com/papers/#robust-composition)_.
+
+To prevent tampering, use [harden](https://github.com/endojs/endo/blob/master/packages/ses/README.md#harden), which is a deep form of [Object.freeze](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/freeze).
+
+<<< @/snippets/test-hardened-js.js#defensiveMaker
+
+Any attempt to modify the properties of a hardened object throws:
+
+<<< @/snippets/test-hardened-js.js#thwarted
+
+`harden()` should be called on all objects that will be transferred
+across a trust boundary. It's important to `harden()` an object before exposing the object by returning it or passing it to some other function.
+
+::: tip harden(), classes, and details
+Note that hardening a class instance also hardens the class.
+For more details, see [harden API in the `ses` package](https://github.com/endojs/endo/blob/master/packages/ses/README.md#harden)
+:::
+## Objects with state
+
+Now let's review the `makeCounter` example:
+
+<<< @/snippets/test-hardened-js.js#counterAnimation
+
+Each call to `makeCounter` creates a new encapsulated `count` variable
+along with `incr` and `decr` functions. The `incr` and `decr` functions [close over](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Closures) the `count` variable.
```

Comment:
I've always hated the "close over" terminology. If you have not heard it before, you cannot correctly guess its meaning. The linked to MDN explanation does not use that term. I like "capture", and the MDN document (thankfully!) does use that term.